Kudu.Core/Jobs/BaseJobRunner.cs

```diff
@@ -102,10 +89,14 @@ private void CacheJobBinaries(JobBase job, IJobLogger logger)
             {
                 try
                 {
-                    int currentHash = CalculateHashForJob(JobBinariesPath);
-                    int lastHash = CalculateHashForJob(WorkingDirectory);
-
-                    if (lastHash == currentHash)
+                    // If the newest file in the WebJobs folder is newer than the newest file in the cache,
+                    // treat the cache as invalid.
+                    // Note that we don't want to invalidate the cache if files are added in the cache. Also, we
+                    // update the .exe.config file, and we don't want that to invalidate the cache.
+                    DateTime lastWebJobsWriteTime = FileSystemHelpers.GetFolderLatestWriteTimeUtc(JobBinariesPath);
+                    DateTime lastWorkingDirectoryWriteTime = FileSystemHelpers.GetFolderLatestWriteTimeUtc(WorkingDirectory);
+
+                    if (lastWebJobsWriteTime <= lastWorkingDirectoryWriteTime)
```

Comment:
There is no smart copy. It kills the whole tmp folder and creates a new one. I'd say smart copy is conceivable (e.g. with KuduSync, manifest, ...), but I'm not sure it is worth the complexity. As long as we fixed the logic so it correctly uses the cache when it should, there shouldn't be too much file copying going on.